src/state/slices/opportunitiesSlice/opportunitiesSlice.ts

```diff
@@ -0,0 +1,117 @@
+import { createSlice } from '@reduxjs/toolkit'
+import { createApi } from '@reduxjs/toolkit/query/react'
+import type { AccountId, AssetId } from '@shapeshiftoss/caip'
+import type { DefiType } from 'features/defi/contexts/DefiManagerProvider/DefiCommon'
+import { BASE_RTK_CREATE_API_CONFIG } from 'state/apis/const'
+
+export const initialState: OpportunitiesState = {
+  lp: {
+    byAccountId: {},
+    byId: {},
+    ids: [],
+  },
+  farming: {
+    byAccountId: {},
+    byId: {},
+    ids: [],
+  },
+  userLp: {
+    byId: {},
+    ids: [],
+  },
+  userFarming: {
+    byId: {},
+    ids: [],
+  },
+}
+
+// The contract info, also referred to as "metadata" - doesn't contain any user-specific data
+type Opportunity = {
+  // The LP token AssetId
+  assetId: AssetId
+  provider: string
+  tvl: string
+  apy: string
+  type: DefiType
+}
+
+type LpOpportunity = Opportunity & {
+  // A tuple of the underlying assets this LP represents
+  underlyingAssetIds: [AssetId, AssetId]
+}
+
+type FarmingOpportunity = Opportunity
+
+// User-specific values for this opportunity
+type UserLpOpportunity = LpOpportunity & {
+  // The amount of held LP tokens
+  cryptoAmount: string
+  underlyingFoxCryptoBaseUnitAmount: string
```

Comment:
this should probably be a string tuple in the order of underlyingAssetIds, and not refer to any specific asset here so this is extensible.

we should also establish an ordering of those asset ids. given they're static values, we should probably enforce lexicographical order for a pair of LP tokens.

Reply:
This is already in the base opportunity:
https://github.com/shapeshift/web/blob/5d6766d0eac5a0e970ae1bf5b8d9b3d9b038199d/src/state/slices/opportunitiesSlice/opportunitiesSlice.ts#L40

What about something generic along the lines of `underlyingAssetACryptoBaseUnitAmount` and `underlyingAssetBCryptoBaseUnitAmount`? Obviously, this would be tied to the order of the underlying AssetIds tuple, so we would have to be able to keep it generic enough in the slice (`[AssetId, AssetId]`), while enforcing that a specific LP is a narrowed IDs subset of said tuple?

Reply:
`underlyingAssetAmounts = [string, string]` will generalize across all LPs?